File: latexdraw-core/net.sf.latexdraw/src/main/net/sf/latexdraw/actions/shape/PasteShapes.scala
package net.sf.latexdraw.actions.shape

import java.util.ArrayList
import org.malai.action.Action
import org.malai.undo.Undoable
import net.sf.latexdraw.actions.DrawingAction
import net.sf.latexdraw.actions.Modifying
import net.sf.latexdraw.glib.models.ShapeFactory
import net.sf.latexdraw.glib.models.interfaces.shape.IShape
import net.sf.latexdraw.glib.views.MagneticGrid
import net.sf.latexdraw.util.LangTool

/**
 * This action pastes the copied or cut shapes.<br>
 * <br>
 * This file is part of LaTeXDraw.<br>
 * Copyright (c) 2005-2014 Arnaud BLOUIN<br>
 * <br>
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * <br>
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.<br>
 * <br>
 * 2012-04-20<br>
 * @author Arnaud BLOUIN
 * @since 3.0
 */
class PasteShapes extends Action with DrawingAction with Undoable with Modifying {
	/** The cut or copy action. */
	var _copy : CopyShapes = _

	/** The magnetic grid to use. */
	var _grid : MagneticGrid = _

	var pastedShapes : List[IShape] = Nil

	override def canDo = _copy!=null && _drawing.isDefined //&& _grid!=null FIXME


	override def isRegisterable = true


	override def doActionBody() {
		val dr = _drawing.get
		// While pasting cut shapes, the first paste must be at the same position that the original shapes.
		// But for pasting after just copying, a initial gap must be used.
		if(!_copy.isInstanceOf[CutShapes])
			_copy.nbTimeCopied+=1

		val gapPaste = 10//if(_grid.isMagnetic) _grid.getGridSpacing else 10 // FIXME
		val gap = _copy.nbTimeCopied*gapPaste

		_copy.copiedShapes.forEach{shape =>
			val sh = ShapeFactory.duplicate(shape)
			pastedShapes = pastedShapes :+ sh
			sh.translate(gap, gap)
			dr.addShape(sh)
		}

		if(_copy.isInstanceOf[CutShapes])
			_copy.nbTimeCopied+=1

		println("paster")
		dr.setModified(true)
	}


	override def undo() {
		var i = 0
		val nbShapes = _copy.copiedShapes.size
		val dr = _drawing.get

		while(i<nbShapes && !dr.isEmpty) {
			dr.removeShape(dr.size-1)
			i+=1
		}

		_copy.nbTimeCopied-=1
		dr.setModified(true)
	}


	override def redo() {
		val dr = _drawing.get
		if(!_copy.isInstanceOf[CutShapes])
			_copy.nbTimeCopied+=1

		pastedShapes.foreach{sh => dr.addShape(sh)}

		if(_copy.isInstanceOf[CutShapes])
			_copy.nbTimeCopied+=1

		dr.setModified(true)
	}


	override def getUndoName = LangTool.INSTANCE.getBundle.getString("LaTeXDrawFrame.43")


	override def followingActions = {
		val list = new ArrayList[Action]
		val selectAction = new SelectShapes
		selectAction.setDrawing(_drawing.get)
		pastedShapes.foreach{sh => selectAction.addShape(sh)}
		list.add(selectAction)
		list
	}


	/**
	 * @param copy the copy to set.
	 */
	def setCopy(copy : CopyShapes) {
		_copy = copy
	}

	def copy = _copy

	/** Sets the magnetic grisd to use. */
	def setGrid(grid:MagneticGrid) {
		_grid = grid
	}
}

File: latexdraw-core/net.sf.latexdraw/src/main/net/sf/latexdraw/actions/shape/PasteShapes.java
/*
 * This file is part of LaTeXDraw.
 * Copyright (c) 2005-2017 Arnaud BLOUIN
 * LaTeXDraw is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option) any later version.
 * LaTeXDraw is distributed without any warranty; without even the implied
 * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 */
package net.sf.latexdraw.actions.shape;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import net.sf.latexdraw.actions.DrawingActionImpl;
import net.sf.latexdraw.actions.Modifying;
import net.sf.latexdraw.models.ShapeFactory;
import net.sf.latexdraw.models.interfaces.shape.IShape;
import net.sf.latexdraw.util.LangTool;
import net.sf.latexdraw.view.MagneticGrid;
import org.malai.action.Action;
import org.malai.undo.Undoable;

/**
 * This action pastes the copied or cut shapes.
 * @author Arnaud Blouin
 */
public class PasteShapes extends DrawingActionImpl implements Undoable, Modifying {
	/** The cut or copy action. */
	private CopyShapes copy;

	/** The magnetic grid to use. */
	private MagneticGrid grid;

	private final List<IShape> pastedShapes;


	public PasteShapes() {
		super();
		pastedShapes = new ArrayList<>();
	}

	@Override
	public boolean canDo() {
		return super.canDo() && copy != null && grid != null;
	}

	@Override
	public boolean isRegisterable() {
		return true;
	}

	@Override
	public void doActionBody() {
		drawing.ifPresent(dr -> {
			// While pasting cut shapes, the first paste must be at the same position that the original shapes.
			// But for pasting after just copying, a initial gap must be used.
			if(!(copy instanceof CutShapes)) {
				copy.nbTimeCopied++;
			}

			final int gapPaste = grid.isMagnetic() ? grid.getGridSpacing() : 10;
			final int gap = copy.nbTimeCopied * gapPaste;

			copy.copiedShapes.forEach(shape -> {
				final IShape sh = ShapeFactory.INST.duplicate(shape);
				pastedShapes.add(sh);
				sh.translate(gap, gap);
				dr.addShape(sh);
			});

			if(copy instanceof CutShapes) {
				copy.nbTimeCopied++;
			}

			dr.setModified(true);
		});
	}

	@Override
	public void undo() {
		drawing.ifPresent(dr -> {
			int i = 0;
			final int nbShapes = copy.copiedShapes.size();

			while(i < nbShapes && !dr.isEmpty()) {
				dr.removeShape(dr.size() - 1);
				i++;
			}

			copy.nbTimeCopied--;
			dr.setModified(true);
		});
	}

	@Override
	public void redo() {
		drawing.ifPresent(dr -> {
			if(!(copy instanceof CutShapes)) {
				copy.nbTimeCopied++;
			}

			pastedShapes.forEach(dr::addShape);

			if(copy instanceof CutShapes) {
				copy.nbTimeCopied++;
			}

			dr.setModified(true);
		});
	}

	@Override
	public String getUndoName() {
		return LangTool.INSTANCE.getBundle().getString("LaTeXDrawFrame.43");
	}

	@Override
	public List<Action> followingActions() {
		if(drawing.isPresent()) return Collections.emptyList();

		final List<Action> list = new ArrayList<>();
		final SelectShapes selectAction = new SelectShapes();
		selectAction.setDrawing(drawing.get());
		pastedShapes.forEach(selectAction::addShape);
		list.add(selectAction);
		return list;
	}

	public void setCopy(final CopyShapes copysh) {
		copy = copysh;
	}

	public CopyShapes getCopy() {
		return copy;
	}

	/** Sets the magnetic grid to use. */
	public void setGrid(final MagneticGrid gr) {
		grid = gr;
	}
}
